### ISSUE TITLE
Missing API templates in Aristotle 2.1.2

### ISSUE THREAD

**Issue by rknowles9** [aristotle-mdr/aristotle-metadata-registry#879]
I installed Aristotle:

> => pip install cffi        #next install step fails if I don't do this first.
> => pip install Aristotle-metadata-registry
> => rehash
> => aristotle-installer
> #(install all extensions)
> => python mdr/manage.py runserver

Email from Sam last week led me to believe that these links would work:

> http://localhost:8000/api/
> http://localhost:8000/api/v3/
> http://localhost:8000/api/v3/schemas/
> http://localhost:8000/api/graphql/

based on links to https://registry.aristotlemetadata.com/ that he sent. The only one working for me is http://localhost:8000/api/v3/. 
http://localhost:8000/api/ gives "NoReverseMatch at /api/", 
http://localhost:8000/api/v3/schemas/ gives "TemplateDoesNotExist at /api/v3/schemas/", and 
http://localhost:8000/api/graphql/ gives "TemplateDoesNotExist at /api/graphql/api".

What am I missing here? Are these correct entry points? Do I need additional installation steps? I ran "python mdr/manage.py help" but didn't see any commands that seemed particularly API or graphql.

Thanks,
Dick Knowles

**Comment by harrywhite4** [comment]
So those are the correct urls. 

The NoReveseMatch error probably means the api's urls are not in urls.py. Both the api and graphql urls need to be included there and the packages listed in installed_apps in settings.py. The easy installer does not currently setup the graphql api.

Does it say which templates are not found when you get the TemplateDoesNotExist errors?

**Comment by rknowles9** [comment]
Here's urls.py from my installed metadata registry:
```
from django.conf.urls import include, url

# Below are the 'recommended' locations for the paths to aristotle extensions.
# At this point the 'glossary' extension *requires* being in the root at `/glossary/`.
urlpatterns = (
    url(r'^', include('aristotle_mdr.urls')),
    url(r'^browse/', include('aristotle_mdr.contrib.browse.urls')),
    url(r'^help/', include('aristotle_mdr.contrib.help.urls', app_name="aristotle_help", namespace="aristotle_help")),
    url(r'^', include('aristotle_mdr.contrib.identifiers.urls', app_name="aristotle_mdr_identifiers", namespace="aristotle_identifiers")),
    #url(r'^', include('aristotle_mdr.contrib.links.urls', app_name="aristotle_mdr_links", namespace="aristotle_mdr_links")
    url(r'^dse/', include('aristotle_dse.urls',app_name="aristotle_dse",namespace="aristotle_dse")),
    url(r'^glossary/', include('aristotle_glossary.urls',app_name="aristotle_glossary",namespace="glossary")),
    url(r'^api/', include('aristotle_mdr_api.urls',app_name="aristotle_mdr_api",namespace="aristotle_mdr_api")),
    url(r'^api/graphql/', include('aristotle_mdr_graphql.urls', namespace="aristotle_graphql")),
    url(r'^mallard/', include('mallard_qr.urls',app_name="mallard_qr",namespace="mallard_qr")),
    url(r'^comet/', include('comet.urls', namespace='comet')),
    )

````
and part of settings.py:
```
INSTALLED_APPS = (
     'aristotle_dse', # Additional models for describing datasets - https://github.com/aristotle-mdr/aristotle-dataset-extensions
     'aristotle_glossary', # Model for managing and inserting glossary content - https://github.com/aristotle-mdr/aristotle-glossary
     'aristotle_mdr_api', # JSON API for programmatic access to content
     'rest_framework', # Needed for the above
     'aristotle_mdr_graphql',
     'comet',
     'mallard_qr',
) + INSTALLED_APPS # Installs the required apps to run aristotle.

```
Looks like they have both api and graphql. Here's more of the error for http://localhost:8000/api/:
```
NoReverseMatch at /api/

Reverse for 'aristotle_api_root' not found. 'aristotle_api_root' is not a valid view function or pattern name.

Request Method: 	GET
Request URL: 	http://localhost:8000/api/
Django Version: 	1.11.15
Exception Type: 	NoReverseMatch
Exception Value: 	

Reverse for 'aristotle_api_root' not found. 'aristotle_api_root' is not a valid view function or pattern name.

Exception Location: 	/Python/ri22432/08-03/venv/lib64/python3.4/site-packages/django/urls/resolvers.py in _reverse_with_prefix, line 497
Python Executable: 	/Python/ri22432/08-03/venv/bin/python
Python Version: 	3.4.8
Python Path: 	

['/Python/ri22432/08-03/mdr',
 '/usr/lib64/python34.zip',
 '/usr/lib64/python3.4',
 '/usr/lib64/python3.4/plat-linux',
 '/usr/lib64/python3.4/lib-dynload',
 '/Python/ri22432/08-03/venv/lib64/python3.4/site-packages',
 '/Python/ri22432/08-03/venv/lib/python3.4/site-packages']

Server time: 	Mon, 6 Aug 2018 15:54:39 +0000
```

As for the TemplateDoesNotExist errors, here's more of the error for http://localhost:8000/api/graphql/ :
```

TemplateDoesNotExist at /api/graphql/api

graphene/graphiql.html

Request Method: 	GET
Request URL: 	http://localhost:8000/api/graphql/api?noexplorer
Django Version: 	1.11.15
Exception Type: 	TemplateDoesNotExist
Exception Value: 	

graphene/graphiql.html

Exception Location: 	/Python/ri22432/08-03/venv/lib64/python3.4/site-packages/django/template/loader.py in get_template, line 25
Python Executable: 	/Python/ri22432/08-03/venv/bin/python
Python Version: 	3.4.8
Python Path: 	

['/Python/ri22432/08-03/mdr',
 '/usr/lib64/python34.zip',
 '/usr/lib64/python3.4',
 '/usr/lib64/python3.4/plat-linux',
 '/usr/lib64/python3.4/lib-dynload',
 '/Python/ri22432/08-03/venv/lib64/python3.4/site-packages',
 '/Python/ri22432/08-03/venv/lib/python3.4/site-packages']

Server time: 	Mon, 6 Aug 2018 15:49:38 +0000
```
And here's the error for http://localhost:8000/api/v3/schemas/ :
```
TemplateDoesNotExist at /api/v3/schemas/

rest_framework_swagger/index.html

Request Method: 	GET
Request URL: 	http://localhost:8000/api/v3/schemas/
Django Version: 	1.11.15
Exception Type: 	TemplateDoesNotExist
Exception Value: 	

rest_framework_swagger/index.html

Exception Location: 	/Python/ri22432/08-03/venv/lib64/python3.4/site-packages/django/template/loader.py in get_template, line 25
Python Executable: 	/Python/ri22432/08-03/venv/bin/python
Python Version: 	3.4.8
Python Path: 	

['/Python/ri22432/08-03/mdr',
 '/usr/lib64/python34.zip',
 '/usr/lib64/python3.4',
 '/usr/lib64/python3.4/plat-linux',
 '/usr/lib64/python3.4/lib-dynload',
 '/Python/ri22432/08-03/venv/lib64/python3.4/site-packages',
 '/Python/ri22432/08-03/venv/lib/python3.4/site-packages']

Server time: 	Mon, 6 Aug 2018 15:59:01 +0000
```

I'll comment again if I find anything else today.

Thanks,
Dick

**Comment by rknowles9** [comment]
I was able to get past both TemplateDoesNotExist  errors by adding to INSTALLED_APPS:
```
INSTALLED_APPS = (
     'aristotle_dse', # Additional models for describing datasets - https://github.com/aristotle-mdr/aristotle-dataset-extensions
     'aristotle_glossary', # Model for managing and inserting glossary content - https://github.com/aristotle-mdr/aristotle-glossary
     'aristotle_mdr_api', # JSON API for programmatic access to content
     'rest_framework', # Needed for the above
     'aristotle_mdr_graphql',
     'comet',
     'mallard_qr',

    'graphene_django',                   #ADDED
    'rest_framework_swagger',            #ADDED
) + INSTALLED_APPS # Installs the required apps to run aristotle.
```
I can continue working now, but should this be necessary?

**Comment by harrywhite4** [comment]
Yes those are definitely necessary, you may also need to add the other apps listed in required_apps here https://github.com/aristotle-mdr/aristotle-metadata-registry/blob/develop/python/aristotle-mdr-api/aristotle_mdr_api/settings.py

You should also add the following 2 lines to your settings.py for the api configuration if they are not already there

``
from aristotle_mdr_api.settings import (
    SERIALIZATION_MODULES,
    REST_FRAMEWORK
)
``
``
from aristotle_mdr_graphql.required_settings import *
``

Also we don't use a namespace for the api so if you remove `namespace="aristotle_mdr_api"` from your urls.py that should clear up the reverse match error.

The setup for the api's isn't the most straight forward process at the moment. So we are looking into documenting this better and possibly adding it to the easy-installer

**Comment by rknowles9** [comment]
Thanks Harry. The missing links all seem to work now. The changes are a combination of what you told me here, email from Sam, and my own discoveries. I want to point that out so that work on #881 will include all necessary steps.

For installation I found 
`=> pip install cffi
`
was necessary. Maybe this just needs to be in requirements somewhere?

The actual install command
`=> aristotle-installer`
was in email from Sam to me. I didn't find this anywhere else.

For INSTALLED_APPS I used the list you pointed out in your previous response, but I still had to add 
`'graphene_django',`

Thanks again. I'm good to close this once you get this last response.
